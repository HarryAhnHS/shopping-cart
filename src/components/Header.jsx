import { Link } from "react-router-dom";
import Wavelength from '/wavelength.png'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBasketShopping } from "@fortawesome/free-solid-svg-icons";

const Header = ({cart, toggleCart}) => {

    function consolidateCart() {
        return cart.reduce((accumulator, item) => accumulator + item.quantity,0,);
    }

    return (
        <div className="px-6 pt-6 text-[#283618]">
            <div className="relative flex flex-col items-center gap-6 sm:flex-row sm:justify-between tracking-tight text-lg">
                <Link to="/" className="cursor-pointer">
                    <div className="w-[200px] sm:w-[240px]">
                        <img src={Wavelength} alt="logo" className="w-full"></img>
                    </div>
                </Link>
                
                <div className="text-xl sm:text-lg sm:mr-24 flex items-center px-6 gap-16 sm:gap-12">
                    <Link className="hover:underline hover:underline-offset-4" to="/">Home</Link>
                    <Link className="hover:underline hover:underline-offset-4" to="/shop">Shop</Link>    
                </div> 

                <button className="absolute right-3 sm:right-6 sm:ml-12 text-[20px]" aria-label="cart button" onClick={() => toggleCart()}>
                        <FontAwesomeIcon icon={faBasketShopping} />
                        <span 
                            className="absolute bottom-3 left-6 bg-red-700 text-white text-[12px] w-[18px] h-[18px] rounded-full flex items-center justify-center"
                            aria-label="cart count"
                        >
                            {consolidateCart()}
                        </span>
                </button>
            </div>
        </div>

    )
}

export default Header;