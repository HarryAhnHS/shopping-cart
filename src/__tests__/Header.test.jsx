import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import "@testing-library/jest-dom";
import { MemoryRouter } from "react-router";
// import userEvent from '@testing-library/user-event'
import { vi } from 'vitest' 
import Header from '../components/Header';

const mockToggleCart = vi.fn();
const mockCart = [{id: 1, title: "MockA", price: 10, images:"", quantity: 5, favorite: false}];

describe('Header', () => {
  it('renders home, shop, and cart links', () => {
    render(
      <MemoryRouter>
        <Header cart={mockCart} toggleCart={mockToggleCart}/>
      </MemoryRouter>
    );
    const homeLink = screen.getByRole("link", { name: /Home/i });
    const shopLink = screen.getByRole("link", { name: /Shop/i });
    const cart = screen.getByRole("button", { name: /Cart/i });

    expect(homeLink).toBeInTheDocument();
    expect(homeLink).toHaveAttribute("href");
    expect(shopLink).toBeInTheDocument();
    expect(shopLink).toHaveAttribute("href");
    expect(cart).toBeInTheDocument();
  });

  it("nav links path should be correct", () => {
    render(
      <MemoryRouter>
        <Header cart={mockCart} toggleCart={mockToggleCart}/>
      </MemoryRouter>
    );
    const homeLink = screen.getByRole("link", { name: /^home$/i });
    const storeLink = screen.getByRole("link", { name: /^shop$/i });
    expect(homeLink).toHaveAttribute("href", "/");
    expect(storeLink).toHaveAttribute("href", "/shop");
  });

  it("should render nav and match snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <Header cart={mockCart} toggleCart={mockToggleCart}/>
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it("should set cart item count to 5", () => {
    render(
      <MemoryRouter>
        <Header cart={mockCart} toggleCart={mockToggleCart}/>
      </MemoryRouter>
    );
    expect(screen.getByRole('button', { name: /5/i })).toBeInTheDocument();
  });
});

