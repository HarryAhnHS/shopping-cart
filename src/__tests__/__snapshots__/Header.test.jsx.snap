// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > should render nav and match snapshot 1`] = `
<div>
  <div>
    <div
      class="flex py-5 bg-red-800 text-white"
    >
      <a
        class="mx-1"
        href="/"
      >
        Home
      </a>
      <a
        class="mx-1"
        href="/shop"
      >
        Shop
      </a>
      <button
        class="mx-1"
      >
        Cart (
        5
        )
      </button>
    </div>
  </div>
</div>
`;
